### PR TITLE
[Feat] migrate to jax

### LINKED ISSUE #59: [Feat] Migrate to JAX

### PR DESCRIPTION
**Pull request by hanson-hschang**

### CODE REVIEW

File: examples/system/hidden_markov_model/__main__.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+    system = HiddenMarkovModel(
+        transition_matrix=jnp.array([[0.7, 0.3], [0.4, 0.6]]),
+        emission_matrix=jnp.array([[0.8, 0.2], [0.2, 0.8]]),
+    )
+
+    print("transition_matrix:\n", system.transition_matrix)
+    print("emission_matrix:\n", system.emission_matrix)
+
+    system = system.with_transition_matrix(
```
Comment (skim0119) [inline examples/system/hidden_markov_model/__main__.py]: Why reuse `system`?

Comment (hanson-hschang) [reply examples/system/hidden_markov_model/__main__.py]: This is to test how to reassign matrix value. Maybe it should be a test instead of an example?

File: src/ss/system/discrete.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+from ._system import System
+
+
+class HiddenMarkovModel(System):
+    _transition_matrix: ProbabilityParameter
+    _emission_matrix: ProbabilityParameter
+    discrete_state_dim: int = eqx.field(static=True)
+    discrete_observation_dim: int = eqx.field(static=True)
+
+    def __init__(
```
Comment (skim0119) [inline src/ss/system/discrete.py]: This feels convoluted. `System` is `eqn.Module` (or python `dataclass`) so having separately overriding `__init__` is just another layer of handling module parameters. Now, there is three location to worry about data in the class: inside `System`, class variables of `class HiddenMarkovModel`, and `__init__`.

It is probably simpler and extendable if one just do:

```py
class HiddenMarkovModel(System):  # basically, dataclass
    transition_matrix: ProbabilityParameter
    emission_matrix: ProbabilityParameter
    ...
```
And this makes me feel like `ProbabilityParameter` is unnecessary wrapper of just Array. Why not just do

```py
type ProbabilityParameter = Array
# ProbabilityParameter: TypeAlias = Array  # below 3.12
class HiddenMarkovModel(System):  # basically, dataclass
    transition_matrix: ProbabilityParameter
    emission_matrix: ProbabilityParameter
    ...
```

Having `class` wrapper around `Array` for operation could be reconsidered. I think just having clear free function (also `jax-pure`) would be syntactically clear.

File: src/ss/system/_system.py [hunk 3]
```diff
@@ -1,167 +1,179 @@
...
+import jax.numpy as jnp
+from jaxtyping import Array, Float, PRNGKeyArray
+
+
+class System(eqx.Module):
+
+    time_step: float = eqx.field(static=True)
+    state_dim: int = eqx.field(static=True)
+    observation_dim: int = eqx.field(static=True)
+    control_dim: int = eqx.field(static=True)
```
Comment (skim0119) [inline src/ss/system/_system.py:22]: `static` necessary?

File: src/ss/system/_system.py [hunk 4]
```diff
@@ -1,167 +1,179 @@
...
 
-    @staticmethod
-    @njit(cache=True)  # type: ignore
-    def _update(
-        array: NDArray,
-        process: NDArray,
-        noise: NDArray,
-    ) -> None:
-        array[:, :] = process + noise
+    def observe(
```
Comment (skim0119) [inline src/ss/system/_system.py:59]: `abstractmethod`?

File: src/ss/system/_system.py [hunk 5]
```diff
@@ -1,167 +1,179 @@
...
-        -------
-        `observation: ArrayLike[float]`
-            The observation vector of systems. Shape of the array is
-            `(batch_size, observation_dim)`.
-        """
-        self._update(
-            self._observation,
-            self._compute_observation_process(),
-            self._compute_observation_noise(),
+    def process(
```
Comment (skim0119) [inline src/ss/system/_system.py:49]: `abstractmethod`?

File: src/ss/system/_system.py [hunk 6]
```diff
@@ -1,167 +1,179 @@
...
-        system: System
-            The created multi-system.
-        """
-        return self.__class__(
-            state_dim=self._state_dim,
-            observation_dim=self._observation_dim,
-            control_dim=self._control_dim,
-            batch_size=batch_size,
-        )
+    def init_state(self) -> Float[Array, "state_dim"]:
```
Comment (skim0119) [inline src/ss/system/_system.py]: `abstractmethod`? Maybe optionally pass RNG key?

File: examples/system/hidden_markov_model/__main__.py (hunk 1 above)
Comment (skim0119) [reply examples/system/hidden_markov_model/__main__.py]: I think this is useful demo. Although, I'm having hard time seeing where is the boundary between the package and user API.

File: src/ss/utility/parameter/probability.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        # TODO: There should be a better way to handle the zero entries in the
+        # softmax output. The current implementation computes the log of the
+        # minimum non-zero value and subtracts a fixed offset to get a "safe"
+        # log value for the zero entries. This is a workaround and may not be
+        # the best approach for all use cases.
+        zero_mask = value == 0
+        log_nonzero_min = jnp.log(jnp.min(jnp.where(zero_mask, jnp.inf, value)))
+        log_zero_value = log_nonzero_min - self.log_zero_offset
+        safe_value = jnp.where(zero_mask, jnp.exp(log_zero_value), value)
+        return jnp.log(safe_value)
```
Comment (skim0119) [inline src/ss/utility/parameter/probability.py:32]:
```suggestion
from functools import partial

class SoftmaxTransformer(eqx.Module):
    """
    A transformer that maps raw values to a probability simplex via the softmax function.
    """

    @staticmethod
    def forward(self, raw_value: Array) -> Array:
        return jax.nn.softmax(raw_value, axis=-1)
    
    @staticmethod
    @partial(jax.jit, static_argnames=("log_zero_offset",))
    def inverse(self, value: Array, log_zero_offset: float=20) -> Array:
        # TODO: There should be a better way to handle the zero entries in the
        # softmax output. The current implementation computes the log of the
        # minimum non-zero value and subtracts a fixed offset to get a "safe"
        # log value for the zero entries. This is a workaround and may not be
        # the best approach for all use cases.
        zero_mask = value == 0
        log_nonzero_min = jnp.log(jnp.min(jnp.where(zero_mask, jnp.inf, value)))
        log_zero_value = log_nonzero_min - log_zero_offset
        safe_value = jnp.where(zero_mask, jnp.exp(log_zero_value), value)
        return jnp.log(safe_value)
```

I think treating them like free function is a bit more cleaner. This is not a stateful operator.

File: src/ss/utility/parameter/_parameter.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    """
+
+    _tensor: Array
+    _transformer: T
+
+    def __init__(self, value: Array, transformer: T) -> None:
+        self._transformer = transformer
+        self._tensor = self._transformer.inverse(value)
+
+    def value(self) -> Array:
```
Comment (skim0119) [inline src/ss/utility/parameter/_parameter.py]: $x.value -> TT^{-1}(x)$ ?

File: src/ss/utility/parameter/_parameter.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+from jaxtyping import Array
+
+
+class Transformer(Protocol):
+    def forward(self, raw_value: Array) -> Array: ...
+    def inverse(self, value: Array) -> Array: ...
+
+T = TypeVar("T", bound=Transformer)
+
+class Parameter(eqx.Module, Generic[T]):
```
Comment (skim0119) [inline src/ss/utility/parameter/_parameter.py:21]: Feels like just a wrapper around `Array`.

```py
_value: Array
value = SomeParameter(_value, operator)

operated_value = value.value()
```

vs

```py
value: Array
operated_value = operator.forward(value)
```

There might be some more complicated usage in the previous branch. Maybe we can discuss later for better structure. I just think, `JAX` code is easier to manage if code is closer to `stateless` and `immutable`, and feels like this is exactly opposite pattern.

File: src/ss/system/_system.py [hunk 10]
```diff
@@ -1,167 +1,179 @@
...
+
+
+class System(eqx.Module):
+
+    time_step: float = eqx.field(static=True)
+    state_dim: int = eqx.field(static=True)
+    observation_dim: int = eqx.field(static=True)
+    control_dim: int = eqx.field(static=True)
+
+    def __check_init__(self) -> None:
```
Comment (skim0119) [inline src/ss/system/_system.py:25]: Nice!

File: src/ss/system/_system.py [hunk 11]
```diff
@@ -1,167 +1,179 @@
...
+    if states.ndim == 1:
+        states = states[:, jnp.newaxis]
+
+    if observations.ndim == 1:
+        observations = observations[:, jnp.newaxis]
+
+    return times, states, observations, controls
+
+
+def batch_simulate(
```
Comment (skim0119) [inline src/ss/system/_system.py]: Would be nice if this is inlined in `simulate`, so that we just have one `simulate` operation. Same for filtering.

Generally prefer `scan(vmap(f))` than `vmap(scan(f))`

File: src/ss/system/_system.py [hunk 12]
```diff
@@ -1,167 +1,179 @@
...
+    The control policy is a function that takes the current time and observation and returns the control input.
+    """
+
+    body = partial(simulate_step, system, control_policy)
+
+    _, (times, states, observations, controls) = jax.lax.scan(
+        body, (initial_time, initial_state), keys
+    )
+
+    if states.ndim == 1:
```
Comment (skim0119) [inline src/ss/system/_system.py]: I think just using states/observations to always have first axis to be `batch_size` would be better (data parallelism). Or, always make the `system` to be single-instance implementation, no batch_size, and let `vmap` + `jax.tree.map` to handle batch (parallel process).

File: src/ss/system/_system.py (hunk 6 above)
Comment (skim0119) [reply src/ss/system/_system.py]: passing key example: 5fb9b23447f4d9b8853d2a0ca56668a26f74f05e

File: src/ss/utility/learning/_process.py [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+        model_trainable, model_static = eqx.partition(model, trainable_mask)
+        optimizer_state = optimizer.init(model_trainable)
+        return cls(model_trainable, model_static, optimizer_state, optimizer)
+
+    @property
+    def model(self) -> Model:
+        return eqx.combine(self.model_trainable, self.model_static)
+
+
+def split_random_key(
```
Comment (skim0119) [inline src/ss/utility/learning/_process.py:48]: Probably can find better place.

File: src/ss/utility/learning/_process.py [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+import jax.numpy as jnp
+import optax
+from jaxtyping import PyTree, Array, PRNGKeyArray
+from tqdm.auto import tqdm
+
+Model = TypeVar("Model", bound=eqx.Module)
+Batch = TypeVar("Batch")
+
+
+@dataclass
```
Comment (skim0119) [inline src/ss/utility/learning/_process.py]:
```suggestion
```

`eqx.Module` includes `dataclass`

File: src/ss/system/_system.py (hunk 12 above)
Comment (skim0119) [reply src/ss/system/_system.py]: d206221

File: src/ss/system/_system.py (hunk 3 above)
Comment (hanson-hschang) [reply src/ss/system/_system.py:22]: yes, it excludes the fields from JAX PyTrees

File: src/ss/utility/parameter/probability.py (hunk 7 above)
Comment (hanson-hschang) [reply src/ss/utility/parameter/probability.py:32]: How about the log_zero_offset? it is something internal. In addition, there might be some other transformation that requires internal state. Then, some transformation will be free function, and some will be object. I think it will create some problem?

File: src/ss/utility/parameter/_parameter.py (hunk 8 above)
Comment (hanson-hschang) [reply src/ss/utility/parameter/_parameter.py]: correct. However, $T^{-1}(x)$ only do it one time at initialization

File: src/ss/utility/parameter/_parameter.py (hunk 9 above)
Comment (hanson-hschang) [reply src/ss/utility/parameter/_parameter.py:21]: @skim0119 Let's discuss this.

I agree Parameter has some internal value, and it is opposite to the JAX pattern. However, internal value might be needed (required) in some cases.

File: src/ss/utility/learning/_process.py (hunk 13 above)
Comment (hanson-hschang) [reply src/ss/utility/learning/_process.py:48]: agree, `src/ss/utility/random.py`?
